File: solutions/Leetcode_36/leetcode_36.cpp
#include <vector>

class Solution1
{
public:
    bool isValidSudoku(std::vector<std::vector<char>>& board)
    {
        int rows[9][9] = {0}, cols[9][9] = {0}, areas[9][9] = {0};
        for (int i = 0; i < 9; i++)
            for (int j = 0; j < 9; j++)
            {
                if (board[i][j] != '.')
                {
                    int num = board[i][j] - '0' - 1;
                    int k = i / 3 * 3 + j / 3; 
                    if (rows[i][num] || cols[j][num] || areas[k][num]) 
                        return false;
                    rows[i][num] = cols[j][num] = areas[k][num] = 1;
                }
            }
        return true;
    }
};

File: solutions/Leetcode_36/leetcode_36.py
from collections import defaultdict
from typing import List


class Solution1:
    def is_valid_sudoku(self, board: List[List[str]]) -> bool:
        rows = defaultdict(list)
        cols = defaultdict(list)
        areas = defaultdict(list)

        for i in range(len(board)):
            for j in range(len(board[0])):
                x = board[i][j]
                if x != '.':
                    hashed_pos = i // 3 * 3 + j // 3
                    if x in rows[i] or x in cols[j] or x in areas[hashed_pos]:
                        return False
                    rows[i].append(x)
                    cols[j].append(x)
                    areas[hashed_pos].append(x)
        return True
